File: scala/src/main/scala/polytyped/sexp_parser.scala
package simplescala_experimentation.polytyped

sealed trait SExp {
  def asString(): String
}
case class SExpAtom(atom: String) extends SExp {
  def asString(): String = atom
}
case class SExpList(list: List[SExp]) extends SExp {
  def asString(): String = {
    "(" + list.map(_.asString).mkString(" ") + ")"
  }
}

object SExpParser {
  type ParseResult[A] = (A, Int)
}

// the parser is intentionally not built on parser combinators
// for performance reasons
class SExpParser(private val input: String) {
  import SExpParser.ParseResult

  // ---BEGIN CONSTRUCTOR---
  private val len = input.length
  // ---END CONSTRUCTOR---

  // the quotes will end up in the result string
  def parseQuote(startPos: Int): Option[ParseResult[String]] = {
    @scala.annotation.tailrec
    def parseFrom(startPos: Int, accum: List[Char], escaped: Boolean): Option[ParseResult[String]] = {
      if (startPos < len) {
        val nextPos = startPos + 1
        input.charAt(startPos) match {
          case '\\' => {
            if (escaped) {
              parseFrom(nextPos, '\\' :: accum, false)
            } else {
              parseFrom(nextPos, accum, true)
            }
          }
          case '"' => {
            val newAccum = '"' :: accum
            if (escaped) {
              parseFrom(nextPos, newAccum, false)
            } else {
              Some((newAccum.reverse.mkString -> nextPos))
            }
          }
          case 'n' if escaped => {
            parseFrom(nextPos, '\n' :: accum, false)
          }
          case o => {
            if (!escaped) {
              parseFrom(nextPos, o :: accum, false)
            } else {
              // escaped some other character
              None
            }
          }
        }
      } else {
        // no terminating quote
        None
      }
    }

    if (startPos < len && input.charAt(startPos) == '"') {
      parseFrom(startPos + 1, List('"'), false)
    } else {
      None
    }
  }

  def parseAtom(startPos: Int): Option[ParseResult[SExpAtom]] = {
    @scala.annotation.tailrec
    def parseFrom(startPos: Int, accum: List[Char]): Option[ParseResult[SExpAtom]] = {
      def done(): Option[ParseResult[SExpAtom]] = {
	if (accum.isEmpty) {
	  None
	} else {
	  Some((SExpAtom(accum.reverse.mkString), startPos))
	}
      }

      if (startPos < len) {
	input.charAt(startPos) match {
	  case '(' | ')' => done
	  case o => {
	    if (Character.isWhitespace(o)) {
	      done
	    } else {
	      parseFrom(startPos + 1, o :: accum)
	    }
	  }
	}
      } else {
	done
      }
    }

    parseQuote(startPos) match {
      case Some((quote, nextPos)) => Some((SExpAtom(quote) -> nextPos))
      case None => parseFrom(startPos, List())
    }
  }

  def parseList(startPos: Int): Option[ParseResult[SExpList]] = {
    @scala.annotation.tailrec
    def parseFrom(startPos: Int, accum: List[SExp]): Option[ParseResult[SExpList]] = {
      if (startPos < len) {
	input.charAt(startPos) match {
	  case '(' => {
	    parseList(startPos) match {
	      case Some((sexp, newPos)) => {
		parseFrom(newPos, sexp :: accum)
	      }
	      case None => None
	    }
	  }
	  case ')' => {
	    Some((SExpList(accum.reverse), startPos + 1))
	  }
	  case o => {
	    if (Character.isWhitespace(o)) {
	      parseFrom(startPos + 1, accum)
	    } else {
	      parseAtom(startPos) match {
		case Some((atom, newPos)) => {
		  parseFrom(newPos, atom :: accum)
		}
		case None => None
	      }
	    }
	  }
	}
      } else {
	None
      }
    }

    if (startPos < len && input.charAt(startPos) == '(') {
      parseFrom(startPos + 1, List())
    } else {
      None
    }
  }

  // returns the first position that doesn't start with whitespace, which
  // might be beyond the start of the string
  @scala.annotation.tailrec
  final def skipOverWhitespace(startPos: Int): Int = {
    if (startPos < len && Character.isWhitespace(input.charAt(startPos))) {
      skipOverWhitespace(startPos + 1)
    } else {
      startPos
    }
  }

  def parseSExp(startPos: Int): Option[ParseResult[SExp]] = {
    parseList(startPos) match {
      case s@Some(_) => s
      case None => parseAtom(startPos)
    }
  }

  def parse(): Option[List[SExp]] = {
    @scala.annotation.tailrec
    def parseFrom(startPos: Int, accum: List[SExp]): Option[List[SExp]] = {
      val first = skipOverWhitespace(startPos)
      if (first < len) {
	parseSExp(first) match {
	  case Some((sexp, newPos)) => {
	    parseFrom(newPos, sexp :: accum)
	  }
	  case None => None
	}
      } else {
	Some(accum.reverse)
      }
    }

    parseFrom(0, List())
  }
}

File: java/simplescala_experimentation/src/main/java/simplescala_experimentation/parser/SExpParser.java
package simplescala_experimentation.parser;

import simplescala_experimentation.util.ImmutableList;
import simplescala_experimentation.util.Cons;
import simplescala_experimentation.util.Nil;
import simplescala_experimentation.util.Pair;

import java.io.File;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.IOException;

public class SExpParser {
    private final String input;
    private final int len;
    
    public SExpParser(final String input) {
        this.input = input;
        len = input.length();
    }

    public int skipOverWhitespace(int pos) {
        while (pos < len && Character.isWhitespace(input.charAt(pos))) {
            pos++;
        }
        return pos;
    } // skipOverWhitespace

    public Pair<String, Integer> parseQuote(final int startPos) throws SExpParseException {
        if (startPos < len && input.charAt(startPos) == '"') {
            int pos = startPos + 1;
            ImmutableList<Character> accum = new Cons<Character>(new Character('"'),
                                                                 new Nil<Character>());
            boolean escaped = false;

            while (pos < len) {
                final char curChar = input.charAt(pos);
                pos++;
                switch (curChar) {
                case '\\':
                    if (escaped) {
                        accum = new Cons<Character>(new Character('\\'), accum);
                        escaped = false;
                    } else {
                        escaped = true;
                    }
                    break;
                case '"':
                    accum = new Cons<Character>(new Character('"'), accum);
                    if (escaped) {
                        escaped = false;
                    } else {
                        return new Pair<String, Integer>(accum.reverse().mkString(),
                                                         new Integer(pos));
                    }
                    break;
                case 'n':
                    if (escaped) {
                        accum = new Cons<Character>(new Character('\n'), accum);
                        escaped = false;
                    } else {
                        accum = new Cons<Character>(new Character('n'), accum);
                    }
                    break;
                default:
                    if (!escaped) {
                        accum = new Cons<Character>(new Character(curChar), accum);
                    } else {
                        throw new SExpParseException("Escaped non-special character: " + curChar);
                    }
                }
            } // while
            
            throw new SExpParseException("No terminating quote");
        } else {
            throw new SExpParseException("No starting quote");
        }
    } // parseQuote

    private class ParseAtomAccum {
        private ImmutableList<Character> accum;
        public ParseAtomAccum() {
            accum = new Nil<Character>();
        }

        public void addChar(final char c) {
            accum = new Cons<Character>(new Character(c), accum);
        }

        public Pair<SExpAtom, Integer> toResult(final int pos) throws SExpParseException {
            if (accum.isEmpty()) {
                throw new SExpParseException("No characters at position: " + pos);
            } else {
                return new Pair<SExpAtom, Integer>(new SExpAtom(accum.reverse().mkString()),
                                                   new Integer(pos));
            }
        }
    } // ParseAtomAccum
    
    public Pair<SExpAtom, Integer> parseAtom(final int startPos) throws SExpParseException {
        try {
            final Pair<String, Integer> pair = parseQuote(startPos);
            return new Pair<SExpAtom, Integer>(new SExpAtom(pair.first),
                                               pair.second);
        } catch (final SExpParseException e) {
            int pos = startPos;
            final ParseAtomAccum accum = new ParseAtomAccum();
            while (pos < len) {
                final char curChar = input.charAt(pos);
                switch (curChar) {
                case '(':
                case ')':
                    return accum.toResult(pos);
                default:
                    if (Character.isWhitespace(curChar)) {
                        return accum.toResult(pos);
                    } else {
                        accum.addChar(curChar);
                        pos++;
                    }
                }
            }

            return accum.toResult(pos);
        }
    } // parseAtom

    public Pair<SExpList, Integer> parseList(final int startPos) throws SExpParseException {
        if (startPos < len && input.charAt(startPos) == '(') {
            int pos = startPos + 1;
            ImmutableList<SExp> accum = new Nil<SExp>();

            while (pos < len) {
                final char curChar = input.charAt(pos);
                switch (curChar) {
                case '(':
                    final Pair<SExpList, Integer> nestedList = parseList(pos);
                    accum = new Cons<SExp>(nestedList.first, accum);
                    pos = nestedList.second.intValue();
                    break;
                case ')':
                    return new Pair<SExpList, Integer>(new SExpList(accum.reverse()),
                                                       new Integer(pos + 1));
                default:
                    if (Character.isWhitespace(curChar)) {
                        pos++;
                    } else {
                        final Pair<SExpAtom, Integer> nestedAtom = parseAtom(pos);
                        accum = new Cons<SExp>(nestedAtom.first, accum);
                        pos = nestedAtom.second.intValue();
                    }
                }
            }

            throw new SExpParseException("No terminating paren");
        } else {
            throw new SExpParseException("No starting paren");
        }
    } // parseList

    public Pair<SExp, Integer> parseSExp(final int startPos) throws SExpParseException {
        try {
            final Pair<SExpList, Integer> listPair = parseList(startPos);
            return new Pair<SExp, Integer>(listPair.first, listPair.second);
        } catch (final SExpParseException e) {
            final Pair<SExpAtom, Integer> atomPair = parseAtom(startPos);
            return new Pair<SExp, Integer>(atomPair.first, atomPair.second);
        }
    } // parseSExp

    public ImmutableList<SExp> parse() throws SExpParseException {
        ImmutableList<SExp> accum = new Nil<SExp>();
        int pos = skipOverWhitespace(0);
        while (pos < len) {
            final Pair<SExp, Integer> pair = parseSExp(pos);
            accum = new Cons<SExp>(pair.first, accum);
            pos = skipOverWhitespace(pair.second.intValue());
        }

        return accum.reverse();
    } // parse

    public static ImmutableList<SExp> parse(final String input) throws SExpParseException {
        return new SExpParser(input).parse();
    }

    public static ImmutableList<String> readLinesFromReaderNoClose(final BufferedReader reader) throws IOException {
        ImmutableList<String> retval = new Nil<String>();
        String line = null;
        while ((line = reader.readLine()) != null) {
            retval = new Cons<String>(line, retval);
        }
        return retval.reverse();
    }

    // closes file at the end
    public static ImmutableList<String> readLinesFromReader(final BufferedReader reader) throws IOException {
        final ImmutableList<String> retval = readLinesFromReaderNoClose(reader);
        reader.close();
        return retval;
    }

    public static ImmutableList<SExp> parse(final File file) throws SExpParseException, IOException {
        return parse(readLinesFromReader(new BufferedReader(new FileReader(file))).mkString("\n"));
    }
}

